src/cacao.ts

```diff
@@ -0,0 +1,111 @@
+import * as dagCbor from '@ipld/dag-cbor'
+import * as multiformats from 'multiformats'
+import * as multihashes from 'multihashes'
+import * as sha256 from '@stablelib/sha256'
+import { SiweMessage } from './siwe'
+
+const DAG_CBOR_CODEC_CODE = 113
+
+export type Header = {
+  t: string
+}
+
+export type Payload = {
+  domain: string
+  iss: string
+  chainId: number
+  aud: string
+  version: string
+  nonce: string
+  iat: number
+  nbf?: number
+  exp?: number
+  statement?: string
+  requestId?: string
+  resources?: Array<string>
+}
+
+export type Signature = {
+  s: string
+}
+
+export type Cacao = {
+  h: Header
+  p: Payload
+  s: Signature
+}
+
+export interface CacaoBlock {
+  cacao: Cacao
+  cid: multiformats.CID
+  bytes: Uint8Array
+}
+
+class CACAO {
+  cacao: Cacao
+  constructor(siwe: string | SiweMessage) {
+    const siweMessage = typeof siwe === 'string' ? new SiweMessage(siwe) : siwe
+    if (!siweMessage.signature) {
+      throw new Error('SIWE Message has not been signed yet')
+    }
+    this.cacao = this.fromSiweMessage(siweMessage)
+  }
+
+  async toCacaoBlock(): Promise<CacaoBlock> {
+    const encodedCacao = dagCbor.encode(this.cacao)
+    const digest = sha256.hash(encodedCacao)
+    const sha256Encoder = (input: Uint8Array) => multihashes.encode(input, 'sha2-256')
+    const hasher = new multiformats.hasher.Hasher('dag-cbor', 113, sha256Encoder)
+    const mhDigest = await hasher.digest(digest)
+    const cid = multiformats.CID.createV1(113, mhDigest)
+    return {
+      cacao: this.cacao,
+      cid: cid,
+      bytes: mhDigest.bytes,
```

Comment:
Would it not be better to return a CAR file here? Or do we need the `cid` to know which is the root?
cc @ukstv

Reply:
I still need to implement the API to convert the `CacaoBlock` to a serialized CAR-representation. That work is tracked in a different ticket, and I need to read up more on CAR before implementing that. DIdn't want to bundle up a bunch of code in a single PR.

Reply:
ok 👍